src/lib/runner.js

```diff
@@ -1,127 +1,206 @@
-import fs from 'fs';
-import os from 'os';
-import path from 'path';
-import { spawn, spawnSync } from 'child_process';
+import fs from 'node:fs/promises';
+import os from 'node:os';
+import path from 'node:path';
+import { spawn } from 'node:child_process';
 
 class Runner {
-  constructor(enableStdOut = false, forwardParentArgs = false) {
-    this.enableStdOut = enableStdOut; // debugging tests
-    this.forwardParentArgs = forwardParentArgs;
-    this.setupFiles = [];
-    this.childProcess = null;
+  #rootDir;
+  #setupFiles = [];
+  #childProcess = null;
+
+  #enableDebug = false;
+  #forwardParentArgs = false;
+
+  // JS strings can be up to 2^53 - 1 in length,
+  // so no need to worry about overflow
+  // https://262.ecma-international.org/16.0/index.html#sec-ecmascript-language-types-string-type
+  #stdOutBuffer = '';
+  #stdErrBuffer = '';
+
+  #stdOutSubscribers = [];
+  #stdErrSubscribers = [];
+
+  constructor(enableDebug = false, forwardParentArgs = false) {
+    this.#enableDebug = enableDebug; // debugging tests
+    this.#forwardParentArgs = forwardParentArgs;
   }
 
-  setup(rootDir, setupFiles = [], options = { create: true }) {
-    this.setupFiles = setupFiles;
+  async setup(rootDir, setupFiles = null, options = { create: true }) {
```

Comment:
I switched the default because elsewhere we were passing `null` and it was overriding the default argument of `[]`, but within this function it appeared as if it was going to always be at least an empty array. Hopefully this makes the possible arguments more clear.